### PR TITLE
Add Bert Language Modeling example

### PR DESCRIPTION
**Pull request by yeandy**
Add Bert Language Modeling example

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by yeandy** [comment]
R: @AnandInguva @tvalentyn

**Comment by yeandy** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/examples/inference/pytorch_bert.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        text
+        | 'AddMask' >> beam.Map(add_mask_to_last_word))
+    text_and_tokenized_text_tuple = (
+        text_and_masked_text_tuple
+        | 'TokenizeSentence' >> beam.Map(tokenize_sentence))
+    text_and_predictions = (
+        text_and_tokenized_text_tuple
+        |
+        'PyTorchRunInference' >> RunInference(model_handler).with_output_types(
+            Tuple[str, PredictionResult])
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/pytorch_bert.py:195]: is this still necessary with the proposed changes to make keyed/non-keyed model handlers as separate classes?

File: sdks/python/apache_beam/examples/inference/pytorch_bert.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      help='Path to the text file containing image names.')
+  parser.add_argument(
+      '--output',
+      dest='output',
+      help='Path where to save output predictions.'
+      ' text file.')
+  parser.add_argument(
+      '--model_state_dict_path',
+      dest='model_state_dict_path',
+      default='/Users/yeandy/Downloads/'
```
[the remark this answers is not in the document]
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_bert.py]: Yeah. Fixed.

File: sdks/python/apache_beam/examples/inference/pytorch_bert.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def tokenize_sentence(
+    text_and_mask: Tuple[str, str]) -> Tuple[str, Dict[str, torch.Tensor]]:
+  text, masked_text = text_and_mask
+  tokenized_sentence = BERT_TOKENIZER.encode_plus(
+      masked_text, return_tensors="pt")
+
+  # Workaround to manually remove batch dim until we have the feature to
+  # add optional batching flag. TODO: Remove once optional batching flag added
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/pytorch_bert.py:59]: is there an Issue tracking this TODO?

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        text_and_tokenized_text_tuple
+        | 'PyTorchRunInference' >> RunInference(
+            KeyedModelHandler(model_handler)).with_output_types(
+                Tuple[str, PredictionResult])
+        | 'ProcessOutput' >> beam.ParDo(PostProcessor()))
+    combined_text = (({
+        'masked_text': text_and_masked_text_tuple,
+        'predicted_text': text_and_predictions
+    })
+                     | 'Merge' >> beam.CoGroupByKey()
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:200]: is this something that can be folded into PostProcess? isn't the masked_text already a part of predicted result?

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        text
+        | 'AddMask' >> beam.Map(add_mask_to_last_word))
+    text_and_tokenized_text_tuple = (
+        text_and_masked_text_tuple
+        | 'TokenizeSentence' >> beam.Map(tokenize_sentence))
+    text_and_predictions = (
+        text_and_tokenized_text_tuple
+        | 'PyTorchRunInference' >> RunInference(
+            KeyedModelHandler(model_handler)).with_output_types(
+                Tuple[str, PredictionResult])
```
Comment (tvalentyn) [inline sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:194]: I don't think we need these hints anymore.

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    combined_text = (({
+        'masked_text': text_and_masked_text_tuple,
+        'predicted_text': text_and_predictions
+    })
+                     | 'Merge' >> beam.CoGroupByKey()
+                     | beam.Map(
+                         lambda x: x[1]['masked_text'][0] + ';' + x[0] + ';' +
+                         x[1]['predicted_text'][0]))
+
+    if known_args.output:
```
[the remark this answers is not in the document]
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:205]: It is there to showcase the ability to write predictions to GCS. For notebook execution, we can take out this`output` part and simply log to the notebook.

[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:205]: If users run the example mostly in a non-interactive way, then if a user omits 'output', and the pipeline runs, there won't be any way to check results, so such execution doesn't provide much value, and `output`  should rather be a mandatory param.

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def parse_known_args(argv):
+  """Parses args for the workflow."""
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--input',
+      dest='input',
+      default='gs://apache-beam-ml/datasets/custom/sentences.txt',
+      help='Path to the text file containing image names.')
```
[the remark this answers is not in the document]
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:87]: Fixed.

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  def process(
+      self, element: Tuple[str, PredictionResult]) -> Iterable[Tuple[str, str]]:
+    text, prediction_result = element
+    inputs = prediction_result.example
+    logits = prediction_result.inference['logits']
+    mask_token_index = (
+        inputs['input_ids'] == BERT_TOKENIZER.mask_token_id).nonzero(
+            as_tuple=True)[0]
+    predicted_token_id = logits[mask_token_index].argmax(axis=-1)
+    decoded_text = BERT_TOKENIZER.decode(predicted_token_id)
```
[the remark this answers is not in the document]
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:76]: Fixed

File: sdks/python/apache_beam/examples/inference/pytorch_bert.py (hunk 3 above)
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_bert.py]: Added.

File: sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  # Workaround to manually remove batch dim until we have the feature to
+  # add optional batching flag. TODO: Remove once optional batching flag added
+  return text, {
+      k: torch.squeeze(v)
+      for k, v in dict(tokenized_sentence).items()
+  }
+
+
+class PostProcessor(beam.DoFn):
```
[the remark this answers is not in the document]
Comment (yeandy) [reply sdks/python/apache_beam/examples/inference/pytorch_language_modeling.py:66]: Added.

**Comment by yeandy** [comment]
Added custom text of sentences. Ran local tests 
```
python -m apache_beam.examples.inference.pytorch_language_modeling \
  --input apache_beam/examples/inference/data/sentences.txt \
  --output predictions.csv \
  --model_state_dict_path BertForMaskedLM.pth

python -m apache_beam.examples.inference.pytorch_language_modeling \
  --output predictions.csv \
  --model_state_dict_path BertForMaskedLM.pth
```

File: sdks/python/apache_beam/examples/inference/data/sentences.txt [hunk 10]
```diff
@@ -0,0 +1,10 @@
+The capital of France is Paris .
```
Comment (AnandInguva) [inline sdks/python/apache_beam/examples/inference/data/sentences.txt:1]: I think we need to add Apache License here

Comment (tvalentyn) [reply sdks/python/apache_beam/examples/inference/data/sentences.txt:1]: We could host it in GCS or pass via `beam.Create([list_of_strings])` if input not specified.

**Comment by yeandy** [comment]
Removed CoGBK for simplicity.

File: sdks/python/apache_beam/ml/inference/pytorch_inference_it_test.py [hunk 11]
```diff
@@ -89,6 +90,41 @@ def test_torch_run_inference_imagenet_mobilenetv2(self):
       filename, prediction = prediction.split(',')
       self.assertEqual(_EXPECTED_OUTPUTS[filename], prediction)
 
+  @pytest.mark.uses_pytorch
+  @pytest.mark.it_postcommit
+  def test_torch_run_inference_bert_for_masked_lm(self):
+    test_pipeline = TestPipeline(is_integration_test=True)
+    # Path to text file containing some sentences
+    file_of_sentences = 'gs://apache-beam-ml/datasets/custom/sentences.txt'  # disable: line-too-long
+    output_file_dir = 'gs://apache-beam-ml/testing/predictions'
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/pytorch_inference_it_test.py:99]: For test output, it's better to use a bucket with a lifecycle configured to leave less clutter behind, for example:

:~$ gsutil lifecycle get gs://temp-storage-for-end-to-end-tests/
{"rule": [{"action": {"type": "Delete"}, "condition": {"age": 14}}]}

Lifecycle may be per bucket (not sure if we can configure it just for `./testing/predictions`), so switching outputs for all tests to gs://temp-storage-for-end-to-end-tests/ may be easiest.

cc: @AnandInguva FYI.

**Review by tvalentyn — APPROVED** [review]

File: sdks/python/apache_beam/ml/inference/pytorch_inference_it_test.py (hunk 11 above)
Comment (yeandy) [reply sdks/python/apache_beam/ml/inference/pytorch_inference_it_test.py:99]: SG, thanks.

**Pull request merged**